### PR TITLE
NURBS curve interpolation functions and small bug fix [nurbs-interp-dev]

### PR DESCRIPTION
**Pull request by JacobLotz**
Second part of splitted PR https://github.com/mfem/mfem/pull/2724 just as PR #3226. Others will follow.

This PR adds curve interpolation function for NURBS curves and fixes a small bug in `void NURBSPatch::DegreeElevate(int dir, int t)`. Furthermore, it adds an example to show the usage of curve interpolation functions.

It adds the following functions: 
- `void KnotVector::FindMaxima(Array<int> &ks, Vector &xi, Vector &u)`: this function finds the location of the maxima of the shapefunctions of the knotvector. This is a good distance for the curve interpolation of NURBS curves with all weights being one: B-splines. 
- `void KnotVector::FindInterpolant(Array<Vector*> &x)`: Finds the interpolant of the curve through the given points `x`. Routine from "The NURBS book" - 2nd ed - Piegl and Tiller.

To do:
- Unsure if the location of the example is correct. It is now in miniapps/nurbs with the other NURBS examples.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3232](https://github.com/mfem/mfem/pull/3232) | @JacobLotz | @tzanio | @v-dobrev + @dylan-copeland | 10/26/22 | 4/29/23 | 5/6/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+   mesh->PrintInfo();
+
+   // Print mesh to file for visualisation
+   VisItDataCollection dc = VisItDataCollection("mesh", mesh);
+   dc.SetPrefixPath("solution");
+   dc.SetCycle(0);
+   dc.SetTime(0.0);
+   dc.Save();
+
+   delete kv_o1, kv;
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:234]: The compiler complains about this. Each pointer should have its own `delete` line.

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+         patch(i,0,1) = (*xy[1])[i]*patch(i,0,2);
+      }
+   }
+
+   // Refinement in curve interpolation direction
+   patch.DegreeElevate(1, order-kv_o1->GetOrder());
+   patch.KnotInsert(1, *kv);
+
+   // 3. Open and write mesh output file
+   string mesh_file;
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:181]:
```suggestion
   string mesh_file("sin-fit.mesh");
```

File: mesh/nurbs.hpp [hunk 3]
```diff
@@ -67,6 +67,9 @@ class KnotVector
    void CalcD2Shape(Vector &grad2, int i, double xi) const
    { CalcDnShape(grad2, 2, i, xi); }
 
+   void FindMaxima(Array<int> &ks,Vector &xi,Vector &u);
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:70]: Most functions in this class have no documentation, but it would be helpful to add documentation at least to the new functions.

File: mesh/nurbs.cpp [hunk 4]
```diff
@@ -341,6 +341,100 @@ void KnotVector::CalcDnShape(Vector &gradn, int n, int i, double xi) const
...
+
+            arg2 = 1;
+            CalcShape ( shape, i, arg2);
+            max2 = shape[d];
+
+            arg = (arg1 + arg2)/2;
+            CalcShape ( shape, i, arg);
+            max = shape[d];
+
+            while ( ( max > max1 ) || (max > max2) )
```
Comment (dylan-copeland) [inline mesh/nurbs.cpp:375]: In the miniapp example, this search takes about 50 iterations. Newton's method, restricted to the unit interval, could solve this much faster. Since it is such a small problem, I suppose it will not hurt performance. So I am fine with leaving this as it is, but you can add the Newton solver if you are interested.

File: mesh/nurbs.cpp [hunk 5]
```diff
@@ -341,6 +341,100 @@ void KnotVector::CalcDnShape(Vector &gradn, int n, int i, double xi) const
...
+               ks[j] = i;
+               xi[j] = arg;
+               u[j]  = getKnotLocation(arg, i+Order);
+            }
+         }
+      }
+   }
+}
+
+// Routine from "The NURBS book" - 2nd ed - Piegl and Tiller
```
Comment (dylan-copeland) [inline mesh/nurbs.cpp:405]: Finding a particular routine in the NURBS book can be time-consuming. Could you please add more details on where to find it? For example, these changes would also be helpful:
+// Routine from "The NURBS book" - 2nd ed - Piegl and Tiller (Algorithm A2.2)
 void KnotVector::CalcShape(Vector &shape, int i, double xi) const

+// Routine from "The NURBS book" - 2nd ed - Piegl and Tiller (Algorithm A2.3)
 void KnotVector::CalcDShape(Vector &grad, int i, double xi) const

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+// Sample runs:  ./nurbs_curveint -uw -n 9
+//               ./nurbs_curveint -nw -n 9
+
+//
+// Description:  This example code demonstrates the use of MFEM to interpolate
+//               a curve in a NURBS patch. We first define a square shaped
+//               NURBS patch. We then interpolate a sine function on the bottom
+//               edge. The results can be viewed in VisIt.
+//
+//               Us use curve interpolation for curves with all weights being 1,
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:12]: Us -> We

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+   double a       = 0.1;
+   int ncp        = 9;
+   int order      = 2;
+   bool ifbspline = true;
+
+   args.AddOption(&l, "-l", "--box-side-length",
+                  "Height and width of the box");
+   args.AddOption(&a, "-a", "--sine-ampl",
+                  "Amplitude of the fitted sine function.");
+   args.AddOption(&a, "-a", "--sine-ampl",
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:65]: The `-a` option is added twice.

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+                  "Amplitude of the fitted sine function.");
+   args.AddOption(&ncp, "-n", "--ncp",
+                  "Number of control points used over four box sides.");
+   args.AddOption(&order, "-o", "--order",
+                  "Order of the NURBSPatch");
+   args.AddOption(&ifbspline, "-uw", "--unit-weight", "-nw",
+                  "--non-unit-weight",
+                  "Use a weight of one only for the NURBS-patches (default) or not: B-splines");
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:73]:
```suggestion
                  "Use unit weights for B-splines (default) or not, for general NURBS");
```

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+   {
+      // We locate the controlpoints at the location of the
+      // maxima of the knotvectors. This works very well
+      // for patches with the weight being one.
+      kv->FindMaxima(i_args,xi_args, u_args);
+
+      for (int i = 0; i < ncp; i++)
+      {
+         (*xy[0])[i] = u_args[i]*l;
+         (*xy[1])[i] = a * sin((*xy[0])[i]/l*2*3.1415);
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:142]: You can use `M_PI` instead of `3.1415`. There is another instance below.

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+   xy[1] = new Vector();
+   Vector xi_args, u_args;
+   Array<int> i_args;
+   xy[0]->SetSize(ncp); xy[1]->SetSize(ncp);
+
+   // Refine direction which has fitting
+   patch.DegreeElevate(0, order-kv_o1->GetOrder());
+   if (!ifbspline)
+   {
+      patch(1,0,2) = sqrt(2)/2;
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:128]: Why should only this one point get the non-unit weight? Depending on the order and number of control points, there can be many points on the bottom of the square. Some documentation explaining the choice of weights would be helpful. Also, the result is strange in the case `nurbs_curveint -nw -n 4 -o 3`.

**Comment by JacobLotz** [comment]
Hi @dylan-copeland, thank you for your comments! I see that the example is done a bit quick and dirty. I am currently out of office and will get back to you on Tuesday.

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 1 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp]: Resolved in [e31bfca](https://github.com/mfem/mfem/pull/3232/commits/e31bfca43bf510dfc9e0cd51398ea06efa9707c6)

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 2 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp]: Resolved in [e31bfca](https://github.com/mfem/mfem/pull/3232/commits/e31bfca43bf510dfc9e0cd51398ea06efa9707c6)

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 6 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp]: Resolved in [e31bfca](https://github.com/mfem/mfem/pull/3232/commits/e31bfca43bf510dfc9e0cd51398ea06efa9707c6)

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 7 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp]: Resolved in [e31bfca](https://github.com/mfem/mfem/pull/3232/commits/e31bfca43bf510dfc9e0cd51398ea06efa9707c6)

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 8 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp]: Resolved in [e31bfca](https://github.com/mfem/mfem/pull/3232/commits/e31bfca43bf510dfc9e0cd51398ea06efa9707c6)

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 9 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp]: Resolved in [e31bfca](https://github.com/mfem/mfem/pull/3232/commits/e31bfca43bf510dfc9e0cd51398ea06efa9707c6)

File: mesh/nurbs.cpp (hunk 5 above)
Comment (JacobLotz) [reply mesh/nurbs.cpp:405]: Source of added algorithms added in [59fbbe0](https://github.com/mfem/mfem/pull/3232/commits/59fbbe08fa936fc269f136e6cd1342ae6827113d)

Comment (JacobLotz) [reply mesh/nurbs.cpp:407]: Other suggestions added as well: [0313c71](https://github.com/mfem/mfem/pull/3232/commits/0313c718d871bc4b3b8803db162d3de018859843)

File: mesh/nurbs.hpp (hunk 3 above)
Comment (JacobLotz) [reply mesh/nurbs.hpp]: Documentation added in [0313c71](https://github.com/mfem/mfem/pull/3232/commits/0313c718d871bc4b3b8803db162d3de018859843)

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 10 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: _Depending on the order and number of control points, there can be many points on the bottom of the square. Some documentation explaining the choice of weights would be helpful.  _
I wrote the code with second order patches in mind and did not realize it could have a purpose for higher order patches. I have improved the code in [25f8deb](https://github.com/mfem/mfem/pull/3232/commits/25f8deb432b4e8364a0609f36323f356e398a940) and [20eb64d](https://github.com/mfem/mfem/pull/3232/commits/20eb64d8cf050879baf67303136f404d9895be19). It contains an explanation for the choice of weights and the application of the weight is improved. I have also improved the curve interpolation procedure and clarified the conversion from carthesian coordinates to homogeneous coordinates.

_Also, the result is strange in the case nurbs_curveint -nw -n 4 -o 3_
This had several reasons. 1) As stated above I have improved the curve interpolation procedure, some part of the procedure made it very messy. 2) Apparantly FindMaxima did not work correctly for `KnotVectors` with one element. For lower orders we just have been lucky it worked out. I have applied a quick fix. But I will have to see if this is the one we need.

Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: And most important. Thank your for pointing it out!

Comment (dylan-copeland) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: Thank you @JacobLotz. This is a very nice example, and the first of its kind in MFEM. The NURBS features in MFEM really needed some examples like this, and I appreciate your contributions. Please update us when you are sure the fix is sufficient, and I will review again.

Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: @dylan-copeland 
I have tried applying Newtons method in finding the maxima of the shapefunctions of the knotvectors as you suggested in an other comment. This was a bit tricky as the Newtons method would find a minimum instead of a maximum sometimes. I think I can push Newton's method in the right direction but it would not increase the elegance of the code in my opinion. 

I think the current approach is better as we are very sure we will find the maximum. I have disturbed the initial guess of both `arg1` and `arg2` with a very small number `1e-16` (very close to machine precision) such that we are able to apply the function to `KnotVector`s with one element. With the initial choice of `arg1` and `arg2` without the disturbance we would also not find exactly 0 or 1 so the disturbance does not influence the results. 

I think we can do the review again.

File: mesh/nurbs.cpp (hunk 4 above)
Comment (JacobLotz) [reply mesh/nurbs.cpp:377]: See the comments in a review below: [link](https://github.com/mfem/mfem/pull/3232#discussion_r1008495962)

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 10 above)
Comment (dylan-copeland) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: Thanks @JacobLotz for the improvements. The example is working well now.

I don't understand the issue that requires perturbing the initial guesses by 1e-16. I set `arg1=0` and `arg2=1` and ran with one element, without any trouble. Is there an example where it fails?

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+   xy[0] = new Vector();
+   xy[1] = new Vector();
+   Vector xi_args, u_args;
+   Array<int> i_args;
+   xy[0]->SetSize(ncp); xy[1]->SetSize(ncp);
+
+   // Refine direction which has fitting
+   if (!ifbspline)
+   {
+      // We we alter the weight for demonstration purposes to a random value. This
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:120]: Typo

File: miniapps/nurbs/nurbs_curveint.cpp [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+   for (int i = 0; i < ncp; i++)
+   {
+      patch(i,0,0) = (*xy[0])[i];
+      patch(i,0,1) = (*xy[1])[i];
+   }
+
+   if (!ifbspline)
+   {
+      // Convert to homogeneous coordinates. FindInterpolant returns
+      // carthesian coordinates.
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_curveint.cpp:156]: Cartesian

File: mesh/nurbs.cpp [hunk 13]
```diff
@@ -341,6 +343,101 @@ void KnotVector::CalcDnShape(Vector &gradn, int n, int i, double xi) const
...
+// Algorithm A9.1 p. 369
+void KnotVector::FindInterpolant(Array<Vector*> &x)
+{
+   int order = GetOrder();
+   int ncp = GetNCP();
+
+   // Find interpolation points
```
Comment (dylan-copeland) [inline mesh/nurbs.cpp:414]: Here, the knotvector coordinates are chosen as the maxima of the basis functions. Once the knotvector coordinates are chosen, a system is solved for each control point. For Algorithm A9.1, doesn't the NURBS book discuss other ways to choose the knotvector coordinates, which are simpler and easier to compute than `FindMaxima`? For example, can't the coordinates be choosen as relative chord lengths? Have you compared `FindMaxima` with other options?

File: mesh/nurbs.hpp [hunk 14]
```diff
@@ -67,6 +67,15 @@ class KnotVector
...
    { CalcDnShape(grad2, 2, i, xi); }
 
+   /** Gives the locations of the maxima of the knotvector in reference space. The
+      function gives the knotspan @a ks, the coordinate in the knotspan @a xi
+      and the coordinate of the maximum in parameter space @a u */
+   void FindMaxima(Array<int> &ks, Vector &xi, Vector &u);
+   /** Global curve interpolation through the points @a x. @a x is an array with the
+      length of the spatial dimension containing vectors with spatial coordinates. The
+      controlpoints of the interpolated curve are given in @a x in the same form.*/
+   void FindInterpolant(Array<Vector*> &x);
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:77]: I would prefer `std::vector<Vector>` over `Array<Vector*>`, but the latter is consistent with the existing code, so we can keep it as is for now. Maybe it is too late to change the API.

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 10 above)
Comment (JacobLotz) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: @dylan-copeland, you are able to run it, but the answer is not correct. You can see it for instance by viewing the result in Visit and and increasing multires control or printing the locations of the maxima. Without the perturbation it is only able to find about halve of the locations of the maxima of the `KnotVector`. A good test would be for instance `./nurbs_curveint -n 6 -o 5`. I have attached two print screens of visit below and the locations of the maxima `xi`.  

In short: This results from evaluating the if statement at line 379 `if (max1 < max2)` where in the case of one element and without the disturbance `max1 = max2 = 0`. This will push the algorithm in the wrong direction such that it only finds about halve of the locations of the maxima.

I hope this clarifies the issue!

**Without the disturbance**
`xi = [2.77556e-17 0.25 0.4 0.5 0.5 1]`

faulty

**With the disturbance**
`xi = [1.27756e-16 0.25 0.4 0.6 0.75 1]`

correct

File: mesh/nurbs.hpp (hunk 14 above)
Comment (JacobLotz) [reply mesh/nurbs.hpp:77]: Good suggestion! I think I do not have any useful arguments to add to this comment.

File: mesh/nurbs.cpp (hunk 13 above)
Comment (JacobLotz) [reply mesh/nurbs.cpp:414]: Good question! I have not compared `FindMaxima` with other methods. From what I understand from the NURBS book there is not necessary a right or wrong answer, but of course this book can be a bit outdated.
We choose for `FindMaxima` for two reasons:
- The spacing between the controlpoints is directly dependend on the spacing between the knots in the `KnotVector`
- We have the idea that inverting the matrix at line 433 of nurbs.cpp in `FindInterpolant` is easier when evaluating the shapefunctions at their maxima. 

You can actually apply this method in combination with chordwise spacing. I have the plan to add a C-mesh around a hydrofoil in another pull request where I apply that. However, I am currently still working on that in the branch [nurbs-auto-knot2edge-dev](https://github.com/mfem/mfem/tree/nurbs-auto-knot2edge-dev).

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 10 above)
Comment (dylan-copeland) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: This does help to clarify the difficulty in solving for the maximum. Finding the maximum in an interval can be tricky, and I see now why this binary search type of algorithm needs to start slightly away from the endpoints, where the basis function can be exactly zero. This algorithm is acceptable for this PR, in my opinion, but maybe in the future it could be improved if someone wants to make it more robust and efficient. As we discussed already, Newton's method could be much faster, but it needs a sufficiently close initial guess. It seems that a small number of initial guesses could be used, perhaps chosen according to which basis function is used. Even better, I wonder whether it is possible to find the maxima analytically, based on the code in `KnotVector::CalcShape`. In any case, such optimizations could be done in the future.

**Review by dylan-copeland — APPROVED** [review]

File: mesh/nurbs.cpp (hunk 13 above)
Comment (IdoAkkerman) [reply mesh/nurbs.cpp:414]: The rows of the matrix add up to 1 due to partition of unity property of the basis. If we choose the collocation points to be the maxima of the shapefunctions, we make the matrix as diagonally dominant as possible (assuming the correct ordering of the points). This seemed (based on my expert intuition :) ) to be favorable in terms conditioning (and proofs?).

File: miniapps/nurbs/nurbs_curveint.cpp (hunk 10 above)
Comment (IdoAkkerman) [reply miniapps/nurbs/nurbs_curveint.cpp:123]: @dylan-copeland I agree. This is not the most elegant solution but works.
Keep in mind that robustness is more important then performance. These routines are typically only run a moderate amount of times when building a mesh.

**Review by IdoAkkerman — APPROVED** [review]
Maximum routine could be more elegant or faster, but serves its purpose here.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks @JacobLotz !

**Comment by dylan-copeland** [comment]
@JacobLotz was there a reason not to add `nurbs_curveint` to `miniapps/nurbs/makefile`? It has to be built individually. If it were in `makefile`, then `make` in the directory `miniapps/nurbs` should build `nurbs_curveint`.

**Comment by JacobLotz** [comment]
@dylan-copeland I assumed that the makefile was created using CMake so I have only updated CMakeLists.text. As I understand now MFEM can be build in two ways and there is there is also a separate makefile. I use CMake to build MFEM so that is problably the reason why I did not notice it.

I wil create a separate pull request to fix these issues.
